test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -278,3 +286,109 @@ UCS_TEST_P(test_memtype_cache, diff_mem_types_diff_bufs_keep_mem) {
 
 INSTANTIATE_TEST_CASE_P(mem_type, test_memtype_cache,
                         ::testing::ValuesIn(mem_buffer::supported_mem_types()));
+
+class test_memtype_cache_deferred_create : public test_memtype_cache {
+protected:
+    virtual void init() {
+        /* do nothing */
+    }
+
+    void test_unknown_region_found(const mem_buffer &b) const {
+        if (b.mem_type() == UCS_MEMORY_TYPE_HOST) {
+            return;
+        }
+
+        test_ptr_found(b.ptr(), b.size(), UCS_MEMORY_TYPE_LAST);
+    }
+};
+
+UCS_TEST_P(test_memtype_cache_deferred_create, allocate_and_update) {
+    const size_t size = 1000000;
+    void *ptr;
+
+    {
+        mem_buffer b(size, GetParam());
+
+        test_memtype_cache::init();
+
+        test_unknown_region_found(b);
+        test_region_not_found(b);
+
+        memtype_cache_update(b.ptr(), b.size(), GetParam());
+        test_region_found(b);
+
+        ptr = b.ptr();
+    }
+
+    /* buffer is released */
+    test_ptr_released(ptr, size);
+    test_ptr_not_found(ptr, size);
+}
+
+UCS_TEST_P(test_memtype_cache_deferred_create, allocate_and_update_portions) {
+    const size_t size = 1000000;
+    void *ptr;
+
+    {
+        mem_buffer b(size, GetParam());
+
+        test_memtype_cache::init();
+
+        test_unknown_region_found(b);
+        test_region_not_found(b);
+
+        /* add two protions of the region that shares one adderss in order
+         * to get them merged in a MemType Cache (the last address of the
+         * 1st protion is eqaul to the first address of the 2nd portion): */
+        /* - add the 1st portion */
+        memtype_cache_update(b.ptr(), b.size() / 2, b.mem_type());
+        test_ptr_found(b.ptr(), b.size() / 2, b.mem_type());
+        /* - add the 2nd portion */
+        memtype_cache_update(UCS_PTR_BYTE_OFFSET(b.ptr(), b.size() / 2 - 1),
```

Comment:
can we unite common code with lookup_merged_region
only difference seems that in once case they overlap in 1 byte?
we can pass "overlap_size" (0 or 1) as a parameter to a common helper function

Reply:
good catch, we can
done